### ISSUE TITLE
Write API example?

### ISSUE THREAD

**Issue by imathews** [googleapis/nodejs-bigquery-storage#222]
I saw that this library was recently updated to support the new Storage Write API — would it be possible to publish a simple example for streaming rows into a table, similar to what we have for the Read API? 

Thanks!

**Comment by mbyrne00** [comment]
Yes - tried to reverse-engineer the latest commit to add the functionality but it's not clear how to specify the write stream type (e.g. `PENDING`), write some rows and batch commit. The underlying api in shows some [java examples](https://cloud.google.com/bigquery/docs/write-api#pending).

Would be great to see a simple example that
1. Creates a stream
2. Writes some records
3. Batch commits

**Comment by mbyrne00** [comment]
Also asked here, just in case: https://stackoverflow.com/questions/69793756/write-rows-to-bigquery-via-nodejs-bigquery-storage-write-api?noredirect=1#comment123374327_69793756

**Comment by rayeeskm** [comment]
Have a look at the example i have given for the question raised in https://stackoverflow.com/questions/69793756/write-rows-to-bigquery-via-nodejs-bigquery-storage-write-api

**Comment by mbyrne00** [comment]
Hey @rayeeskm - thanks very much for the example and this answers our question.

On the API itself, I just had some feedback - happy to raise another issue if that's better:

 - It seems odd that we would need to define the `type` and `mode` enums in our code, and that the library does not just provide these exported somewhere. 
 - Is there a way the lib can abstract away the proto complexity and allow callers to specify a JS object? We haven't had to do this for any other google node APIs but maybe this is different. Be great to avoid the need for generated code if we can, going forward.

Thanks again for the example meantime and this would certainly unblock the path to giving it a go.

**Comment by rayeeskm** [comment]
Hey @mbyrne00, yes those enums are available in the library itself. I have updated the example code how to get that from the library.

On the second point of reducing the complexity in defining the proto schema, am yet to figure if this library has any in built mechanism to do that.

**Comment by gal1419** [comment]
@rayeeskm @mbyrne00 I haven't checked it myself yet, but in case it will save you some time - the `AppendRowsRequest` seems to have a static `fromObject` method that converts a plain JS object into `AppendRowsRequest`;

I'll do some testing and will post back.

`
                        /**
                         * Creates an AppendRowsRequest message from a plain object. Also converts values to their respective internal types.
                         * @param object Plain object
                         * @returns AppendRowsRequest
                         */
                        public static fromObject(object: { [k: string]: any }): google.cloud.bigquery.storage.v1.AppendRowsRequest;
`

**Comment by gal1419** [comment]
Update: the above does not work

**Comment by s1moe2** [comment]
Examples would be awesome!
I'm having a hard time understanding why I keep getting a permissions error, even when following the stackoverflow example mentioned above.
`Permission 'TABLES_UPDATE_DATA' denied on resource 'projects/XXX/datasets/university/tables/students' (or it may not exist).`

I'm executing this locally with a service account key which has the right permissions, according to [this guide](https://cloud.google.com/bigquery/docs/write-api#required_permissions).

I've actually activated the service account key in `gcloud` and inserted via the CLI:
```
> cat st.json 
[{ "id": 1, "name": "homer", "age": 50 }]

> gcloud alpha bq tables insert students --dataset=university --data=st.json
kind: bigquery#tableDataInsertAllResponse
```
![image](https://user-images.githubusercontent.com/36140182/143023906-ecda9e08-d6d4-4a4b-aa9d-248f04854106.png)

**Comment by gal1419** [comment]
@s1moe2 this is the exact error I'm also getting. I verified that my service account is configured properly.

**Comment by rayeeskm** [comment]
https://cloud.google.com/docs/authentication/production#linux-or-macos

@s1moe2 @gal1419

**Comment by s1moe2** [comment]
@rayeeskm This doesn't seem to be the problem. I'm [passing the credentials via environment variable](https://cloud.google.com/docs/authentication/production#passing_variable), which should work as well.

When specifying projectId/keyPath, despite getting the error code 7, the error itself is now different (seems more nested maybe):
```
Error: 7 PERMISSION_DENIED: Permission denied on resource project [object Object]
```
(note that `[object Object]` is already part of the error string message...)

**Comment by abcde090** [comment]
I am also getting this `Error: 7 PERMISSION_DENIED: Permission 'TABLES_UPDATE_DATA' denied on resource`. Have anyone solved this issue already ?

**Comment by gal1419** [comment]
@abcde090 I didn't manage to solve it. I'm starting to look at other alternatives. It seems like a bug to me.
@rayeeskm @mbyrne00 - can you please confirm that this is actually working for you?

@steffnay your comment will be much appreciated, thank you in advance.

**Comment by s1moe2** [comment]
I didn't get this to work either. I'll gladly go back to this and open a bug issue so we can compile all the necessary information.
Didn't do it before because it seemed like some people had this working, so I've been tracking this issue in case it was me doing some dumb thing.

**Comment by rayeeskm** [comment]
@gal1419 Yes it works pretty fine.

**Comment by mbyrne00** [comment]
I didn’t get this working, sorry. I had to proceed with ingestion via GCS

**Comment by s1moe2** [comment]
So... Got it working for a couple of minutes!
Decided to try this on a macbook to rule out an os/env issue since on the previous tests I was on a Linux machine. Created a clean project, used an exact copy of the code from @rayeeskm and it worked:
<img width="318" alt="Screenshot 2021-11-28 at 21 47 45" src="https://user-images.githubusercontent.com/36140182/143787325-3846d889-4a63-4254-a8ff-15475963a615.png">

Then I decided to run with Docker, in an node16-alpine image. Worked:
<img width="920" alt="Screenshot 2021-11-28 at 21 40 42" src="https://user-images.githubusercontent.com/36140182/143787381-98c491ad-afad-46fb-b176-0f3aad7eb407.png">

After this, because what I really needed for my use case would be to use the committed mode, I changed it and ran a test to see what would happen. Surprise: `TABLES_UPDATE_DATA`!

So, I reverted all changes to the original code but the error remains.
I also tested the docker version again (I had not made any more builds: the image was still the only one I had built) and it also has the error:
<img width="1128" alt="Screenshot 2021-11-28 at 21 40 57" src="https://user-images.githubusercontent.com/36140182/143787538-3f3e0bc1-02dd-42b7-ae41-9399c6229e17.png">

And here's proof that those two tests (with local node and docker) worked:
![image](https://user-images.githubusercontent.com/36140182/143787556-358c3b9b-baf8-4f90-921c-a54da57fabef.png)

At this point I'm out of ideas. With this sort of tests/results, could this be related to API usage?
Does this ring a bell for anyone? Is there any other piece of data I could look at that would help understand the problem?

**Comment by steffnay** [comment]
@gal1419 @mbyrne00 @abcde090 I'm looking into this. Can you all let me know whether you are encountering this on the free tier or with a project that has billing?

**Comment by gal1419** [comment]
@steffnay answered in #227.
but for the record - my project has billing configured.
thanks for your help!

**Comment by abcde090** [comment]
I have solve this `Error: 7 PERMISSION_DENIED: Permission 'TABLES_UPDATE_DATA' denied on resource` issue. I am not sure why it happened but I created a new dataset in a different location and then it works. My old dataset is located in `australia-southeast1` and my new dataset is in `US`.  It still throws errors on my old dataset in `australia-southeast1` @gal1419 @steffnay @s1moe2

**Comment by steffnay** [comment]
@abcde090 Ah, thank you! I was able to reproduce last night when I changed location and wanted to inquire about which location you all were using, so your update is very helpful.

**Comment by s1moe2** [comment]
@steffnay did you see my last comment on #227 regarding this?

**Comment by tswast** [comment]
For non-default locations, you need to populate the  `x-goog-request-params` header with the value `write_stream=your_stream_name`.

See: https://github.com/googleapis/python-bigquery-storage/blob/ce63994e08091be06b75d4010f2eefb424dbb356/google/cloud/bigquery_storage_v1/writer.py#L165-L172 in Python

**Comment by oh-wo** [comment]
Hi there @steffnay I've found the same thing; wouldn't work in EU locations but when I tested with USA it worked. 

So I found @tswast python-header example above, hoping that I could use that as a work around enabling us to use this api within the EU in advance of your January patch.  Can spot anything wrong with the following?

```typescript
const writeStream: IWriteStream = {
      // No difference observed when I tried Type.PENDING .
      type: Type.COMMITTED,
};
const [ response ] = await storageClient.createWriteStream(request);
const streamName = response.name;
const headers = {
      'x-goog-request-params': `write_stream=${streamName}`,
};

const stream = storageClient.appendRows({ otherArgs: { headers } });
```

Thanks very much.

**Comment by paujur-authentic** [comment]
hello @steffnay - I see this marked as resolved, but I'm still experiencing an issue and seeing this error:

```
Error: 7 PERMISSION_DENIED: Permission 'TABLES_UPDATE_DATA' denied on resource
```
